### PR TITLE
Kadm dbargs

### LINKED ISSUE #8: db_args in init functions
I'm currently invoking the kadmin shell utility to do some tasks against our kerberos which uses LDAP as a KDC Datastore. To create new principals, I need to specifiy the db_args (-x dn=... in the shell utility). Currently, this API doesn't support this (NULL is always passed as the db_args argument in the init functions).

Would you be interested in merging this functionality and cutting a new gem if I were to add it, or do you no longer support rkerberos?

### PR DESCRIPTION
**Pull request by sonOfRa**
Add db_args and db_princ_args functionality. Closes #8

### CODE REVIEW

**Comment by sonOfRa** [comment]
Any comments/suggestions for changes?

**Comment by sonOfRa** [comment]
I've added tests for the db_princ_args (used when creating principals). There are still no tests for the db_args passed to init, as I don't have any proper way to set up an environment that would require these settings.

**Comment by sonOfRa** [comment]
Any further comments or things I need to do?

**Comment by sonOfRa** [comment]
Hello, it's been another month now. Is there any chance at all this is still getting merged, or is this project abandoned?

**Comment by djberg96** [comment]
@domcleal What do you think?

**Comment by domcleal** [comment]
I've not put much work into this at all unfortunately since I took over maintenance from @djberg96 (who seems to be more on the ball than me now), for which I apologise.  I had planned to, but don't even have a Kerberos infrastructure any more.

It looks like @djberg96's looked over it somewhat, so if you're happy, either you or I can merge it (you've still got collab permission, please feel free to use it).

File: ext/rkerberos/kadm5.c [hunk 1]
```diff
@@ -934,6 +958,75 @@ static VALUE rkadm5_randkey_principal(VALUE self, VALUE v_user){
...
+    default:
+      rb_raise(rb_eTypeError, "Need Single String or Array of Strings for db_args");
+  }
+  return db_args;
+}
+
+/**
+ * Add parsed db-args to principal entry
+ */
+void add_db_args(kadm5_principal_ent_rec* entry, char** db_args){
```
Comment (domcleal) [inline ext/rkerberos/kadm5.c]: missing prototype?

File: ext/rkerberos/kadm5.c [hunk 2]
```diff
@@ -934,6 +958,75 @@ static VALUE rkadm5_randkey_principal(VALUE self, VALUE v_user){
...
+    for(i = 0; db_args[i] != NULL; i++){
+      add_tl_data(&entry->n_tl_data, &entry->tl_data, KRB5_TL_DB_ARGS, strlen(db_args[i]) + 1, (krb5_octet*)db_args[i]);
+    }
+  }
+}
+
+/**
+ * Source code taken from kadmin source code at https://github.com/krb5/krb5/blob/master/src/kadmin/cli/kadmin.c
+ */
+void add_tl_data(krb5_int16 *n_tl_datap, krb5_tl_data **tl_datap,
```
Comment (domcleal) [inline ext/rkerberos/kadm5.c]: ditto

**Comment by djberg96** [comment]
@domcleal How would you feel about adding @sonOfRa to the project, then? Assuming he's up for it.

File: ext/rkerberos/extconf.rb [hunk 3]
```diff
@@ -15,4 +15,10 @@
   raise "kadm5clnt library not found"
 end
 
+if have_header('kdb.h')
+	have_library('libkdb5')
```
Comment (domcleal) [inline ext/rkerberos/extconf.rb]: Nitpick, others are indented by two spaces rather than hard tabs.

**Comment by domcleal** [comment]
> How would you feel about adding @sonOfRa to the project, then? Assuming he's up for it.

Yep, I'd be very happy to.

File: ext/rkerberos/kadm5.c (hunk 1 above)
Comment (djberg96) [inline ext/rkerberos/kadm5.c]: Hm, do we prototype all the functions? I don't remember any more.

Comment (domcleal) [inline ext/rkerberos/kadm5.c]: No, I don't think so generally, I just noticed the warnings on this file when I went to build it.

**Comment by sonOfRa** [comment]
So, a year has gone by, and I'm actually back at this after being gone for a year. If you're still willing, I'd be up for taking over a bit in this project and work on some bugfixes and possibly new features.

I can't promise a lot of activity, as this is only one of several projects I work on, but I can at least fix some bugs and add a few features that I and others might find useful in the long run.

**Comment by domcleal** [comment]
Thanks, that'd be good. I've sent you an invite for collaboration permissions.

**Comment by sonOfRa** [comment]
If anyone wants to look over this again I'd be grateful, otherwise I'll merge this tomorrow at some point in the afternoon (UTC)

**Comment by djberg96** [comment]
@sonOfRa FWIW I eyeballed it and it looks good to me. I guess you'll know for sure when people start using it. :)

**Pull request merged**